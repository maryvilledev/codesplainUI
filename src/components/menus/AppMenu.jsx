import React, { Component, PropTypes } from 'react';

import {
  Avatar,
  IconMenu,
  MenuItem,
  IconButton,
} from 'material-ui';
import ArrowDropDown from 'material-ui/svg-icons/navigation/arrow-drop-down';
import MoreVertIcon from 'material-ui/svg-icons/navigation/more-vert';

import OrgSnippetsMenu from './OrgSnippetsMenu';
import SnippetList from './SnippetList';
import CustomPropTypes from '../../util/custom-prop-types';

const styles = {
  iconButtonContainer: {
    display: 'inline-flex',
    alignItems: 'center',
  },
  iconMenu: {
    cursor: 'pointer',
  },
};

// Return an <Avatar /> of the user's GitHub avatar if a URL is specified, else
// Return a generic menu icon
const makeAppMenuIcon = (avatarUrl) => {
  if (avatarUrl) {
    return (
      <div style={styles.iconButtonContainer}>
        <Avatar
          size={30}
          src={avatarUrl}
          style={styles.avatar}
        />
        <ArrowDropDown color="white" />
      </div>
    );
  }
  return (
    <IconButton>
      <MoreVertIcon color="white" />
    </IconButton>
  );
};

class AppMenu extends Component {
  constructor(props) {
    super(props);
    this.state = {
      iconMenuOpen: false,
    };
    this.handleOnRequestChange = this.handleOnRequestChange.bind(this);
    this.handleSnippetSelected = this.handleSnippetSelected.bind(this);
  }

  handleOnRequestChange(open) {
    this.setState({
      iconMenuOpen: open,
    });
  }

  handleSnippetSelected(snippetOwner, snippetKey) {
    this.setState({
      iconMenuOpen: false,
    });
    this.props.onSnippetSelected(snippetOwner, snippetKey);
  }

  render() {
    const {
      avatarUrl,
      onSignOut,
      orgSnippets,
      username,
      userSnippets,
    } = this.props;
    const {
      iconMenuOpen,
    } = this.state;

    return (
      <div>
        <IconMenu
          anchorOrigin={{ horizontal: 'right', vertical: 'top' }}
          iconButtonElement={makeAppMenuIcon(avatarUrl)}
          onItemTouchTap={this.handleOnItemTouchTap}
          onRequestChange={this.handleOnRequestChange}
          open={iconMenuOpen}
          style={styles.iconMenu}
          targetOrigin={{ horizontal: 'right', vertical: 'top' }}
          useLayerForClickAway
        >
          <SnippetList
            onClick={this.handleSnippetSelected}
            primaryText="My Snippets"
            snippetOwner={username}
            snippetsList={userSnippets}
          />
          <OrgSnippetsMenu
            onClick={this.handleSnippetSelected}
            orgSnippets={orgSnippets}
          />
          <MenuItem
            onClick={onSignOut}
            primaryText="Sign out"
          />
        </IconMenu>
      </div>
    );
  }
}

AppMenu.propTypes = {
  avatarUrl: PropTypes.string,
  onSignOut: PropTypes.func.isRequired,
  onSnippetSelected: PropTypes.func.isRequired,
  orgSnippets: CustomPropTypes.orgSnippets.isRequired,
  username: PropTypes.string.isRequired,
  userSnippets: CustomPropTypes.snippets.isRequired,
};

AppMenu.defaultProps = {
  avatarUrl: '',
};

export default AppMenu;
