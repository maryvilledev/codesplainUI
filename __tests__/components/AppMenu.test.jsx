import React from 'react';
import cookie from 'react-cookie';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import getMuiTheme from 'material-ui/styles/getMuiTheme';

import AppMenu from '../../src/components/menus/AppMenu';

describe('<AppMenu />', () => {
  const muiTheme = getMuiTheme();
  const shallowWithContext = (node) => shallow(node, { context: { muiTheme } });

  afterEach(() => {
    cookie.remove('userAvatarURL');
  });

  describe('snapshot tests', () => {
    it('matches snapshot of when user is not logged in', () => {
      const wrapper = shallowWithContext(
        <AppMenu
          onSignOut={jest.fn()}
        />
      );
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    it('matches snapshot of when user is logged in', () => {
      cookie.save('userAvatarURL', 'https://github.com/octocat.png');
      const wrapper = shallowWithContext(
        <AppMenu
          onSignOut={jest.fn()}
        />
      );
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
  });

  describe('prop: onSignOut', () => {
    it('should be forwarded', () => {
      const onSignOut = jest.fn();
      const wrapper = shallowWithContext(
        <AppMenu
          onSignOut={onSignOut}
        />
      );
      expect(wrapper.find('MenuItem').prop('onClick')).toEqual(onSignOut);
    });
  });
});
