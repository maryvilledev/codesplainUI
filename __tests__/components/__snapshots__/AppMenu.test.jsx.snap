exports[`<AppMenu /> snapshot tests matches snapshot of when user is logged in 1`] = `
<div>
  <IconMenu
    anchorOrigin={
      Object {
        "horizontal": "right",
        "vertical": "top",
      }
    }
    animated={true}
    iconButtonElement={
      <div
        style={
          Object {
            "alignItems": "center",
            "display": "inline-flex",
          }
        }>
        <Avatar
          size={30}
          src="https://wallykazam.com/whozit" />
        <NavigationArrowDropDown
          color="white" />
      </div>
    }
    multiple={false}
    onItemTouchTap={[Function]}
    onKeyboardFocus={[Function]}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onRequestChange={[Function]}
    onTouchTap={[Function]}
    open={false}
    style={
      Object {
        "cursor": "pointer",
      }
    }
    targetOrigin={
      Object {
        "horizontal": "right",
        "vertical": "top",
      }
    }
    touchTapCloseDelay={200}
    useLayerForClickAway={true}>
    <SnippetList
      onClick={[Function]}
      primaryText="My Snippets"
      snippetOwner="username"
      snippetsList={Object {}} />
    <OrgSnippetsMenu
      onClick={[Function]}
      orgSnippets={Object {}} />
    <MenuItem
      anchorOrigin={
        Object {
          "horizontal": "right",
          "vertical": "top",
        }
      }
      checked={false}
      desktop={false}
      disabled={false}
      focusState="none"
      insetChildren={false}
      onClick={[Function]}
      primaryText="Sign out" />
  </IconMenu>
</div>
`;

exports[`<AppMenu /> snapshot tests matches snapshot of when user is not logged in 1`] = `
<div>
  <IconMenu
    anchorOrigin={
      Object {
        "horizontal": "right",
        "vertical": "top",
      }
    }
    animated={true}
    iconButtonElement={
      <div
        style={
          Object {
            "alignItems": "center",
            "display": "inline-flex",
          }
        }>
        <Avatar
          size={30}
          src="https://wallykazam.com/whozit" />
        <NavigationArrowDropDown
          color="white" />
      </div>
    }
    multiple={false}
    onItemTouchTap={[Function]}
    onKeyboardFocus={[Function]}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onRequestChange={[Function]}
    onTouchTap={[Function]}
    open={false}
    style={
      Object {
        "cursor": "pointer",
      }
    }
    targetOrigin={
      Object {
        "horizontal": "right",
        "vertical": "top",
      }
    }
    touchTapCloseDelay={200}
    useLayerForClickAway={true}>
    <SnippetList
      onClick={[Function]}
      primaryText="My Snippets"
      snippetOwner="username"
      snippetsList={Object {}} />
    <OrgSnippetsMenu
      onClick={[Function]}
      orgSnippets={Object {}} />
    <MenuItem
      anchorOrigin={
        Object {
          "horizontal": "right",
          "vertical": "top",
        }
      }
      checked={false}
      desktop={false}
      disabled={false}
      focusState="none"
      insetChildren={false}
      onClick={[Function]}
      primaryText="Sign out" />
  </IconMenu>
</div>
`;
